src/main/java/inputStream/ManipulatorInputStream.java

```diff
@@ -0,0 +1,84 @@
+package inputStream;
+
+import manipulateActions.ManipulateAction;
+import status.Status;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class ManipulatorInputStream extends InputStream {
+
+    private ManipulateAction dataManipulatorAction;
+
+    private byte[] dataBuffer;
+
+    private int dataBufferPos;
+
+    private int lastByteRead;
+
+    private InputStream in;
+
+    public ManipulatorInputStream(InputStream in, ManipulateAction dataManipulatorAction) throws IOException {
+        this.in = in;
+
+        this.dataManipulatorAction = dataManipulatorAction;
+
+        this.dataBufferPos = 0;
+
+        this.lastByteRead = in.read();
+    }
+
+    @Override
+    public synchronized int read() throws IOException {
+        if (dataBufferPos == 0 || dataBufferPos == dataBuffer.length) {
+            if (lastByteRead == -1) {
+                return lastByteRead;
+            }
+            dataBuffer = fillDataBuffer((char) lastByteRead).getBytes();
+            dataBufferPos = 0;
+        }
+        return dataBuffer[dataBufferPos++];
+    }
+
+    public String fillDataBuffer(char data) throws IOException {
+        String dataToFill = getDataToFill(data);
+        if (dataManipulatorAction.getStatus() == Status.DATA_CAN_MANIPULATE
+                && dataManipulatorAction.isFiltersStatusGood()) {
+            return dataManipulatorAction.manipulateDataAction(dataToFill);
```

Comment:
why is this the `inputStream`'s responsibility to check this?
an API shouldn't require calling more than one method to perform an action if all it depends on is itself